Allow to make guests moderators

nickvergessen opened the pull request:
* Introduces a new user type (6) GUEST_MODERATOR
* GMs can do everything moderators can do apart from:
  * Adding users (endpoint in server is not public)
  * Make a room private/public (would kill them too)
* GM can be assigned via the UI

Fix #1022 

---

### ToDo
- [ ] Show participant list to guest moderators

fancycode commented on lib/Room.php:
```diff
@@ -484,6 +484,29 @@ public function setParticipantType($participant, $participantType) {
+			->andWhere($query->expr()->eq('session_id', $query->createNamedParameter($participant->getSessionId())));
+		$query->execute();
+
+		$this->dispatcher->dispatch(self::class . '::postSetParticipantTypeBySession', new GenericEvent($this, [
```
Could you please add a listener to notifiy the standalone backend server that the participant has been modified? You can use this as base: https://github.com/nextcloud/spreed/blob/c1d4482d507e44929037b3495cccd63fa5ba0d38/lib/AppInfo/Application.php#L186

nickvergessen replied:
done

fancycode commented on lib/Controller/RoomController.php:
```diff
@@ -525,14 +525,19 @@ public function removeFromFavorites($token) {
 	public function renameRoom($token, $roomName) {
 		try {
```
The same `try { ... }` block is in multiple functions now, can this somehow be refactored in a separate function to avoid code duplication?

Maybe a new decorator similar to `@NoAdminRequired`, e.g. `@TalkModeratorRequired` that performs the code and checks for `hasModeratorPermissions()`.

nickvergessen replied:
yeah, something for later.
I'm not sure if apps can already register middleware annotations.

nickvergessen replied:
But we would then also get the room twice etc, which is kind of dislike about the annotations.

nickvergessen commented:
@danxuliu any idea how we can only show the participant list for guest moderators (not all guests)? See last commit

Peterede commented on lib/Controller/RoomController.php:
```diff
@@ -981,10 +1053,19 @@ public function demoteModerator($token, $participant) {
+		return $this->demoteGuestToModerator($room, $sessionId);
+	}
+
+	protected function demoteUserToModerator(Room $room, string $participant): DataResponse {
```
The name of this function should be changed to indicate that it is removing moderator status from the user not adding it.  demoteUserFromModerator would be a better name.

Peterede commented on lib/Controller/RoomController.php:
```diff
@@ -999,4 +1080,20 @@ public function demoteModerator($token, $participant) {
 		return new DataResponse();
 	}
+
+	protected function demoteGuestToModerator(Room $room, string $sessionId): DataResponse {
```
Similar issue for this function, demoteGuestFromModerator would be a better name.

Peterede commented:
To show the participant list for guest moderators and not all guests it should be not added by default and then when joining if the guest is a moderator call:

OCA.SpreedMe.app._showParticipantList();
OCA.SpreedMe.app._participantsView.setRoom(OCA.SpreedMe.app.activeRoom);

nickvergessen commented:
#1215 is in @danxuliu 
Please rebase here :sunglasses:

danxuliu commented:
I have rebased the branch and fixed some issues:
- Fix removing a participant (`@NoAdminRequired` changed to `@PublicPage`)
- Fix demoting as a moderator (`$this->userId === $participant` in `demoteModerator()` is always true as a guest moderator; the check was moved to `demoteUserToModerator()` and `demoteGuestToModerator()`)

I have squashed in the first commit the commits that fixed code added in it; I have added some other commits with changes to it, but not squashed them (yet) just in case the changes are not desired so they can be easily removed (the rationale for the changes is in the commit message).

Other issues still remain, though.

The avatars in the chat are not properly shown when viewed as a guest user; it is not related to the changes in this pull request, so it will be fixed in a different one.

@nickvergessen There are no system messages when promoting/demoting a guest. Is this by design or simply missing?

@Peterede I agree that `demoteUserToModerator` and `demoteGuestToModerator` should be called `demoteUserFromModerator` and `demoteGuestFromModerator` instead, but according to GitHub you marked that as resolved (even if the old names are still used). What am I missing? :-)

Also, @Peterede, could you provide some steps that show why the change in _js/views/tabview.js_ is needed? I have been trying to trigger the error (without that commit, of course :-P ) and I have not been able to; I guess it was just caused by some of the WIP code and it no longer applies, so I have removed the commit (although I have a reference to it just in case it has to be added back).

nickvergessen commented:
I removed the two commits as discussed.

I fixed the name to demote*FromModerator (copy paste mistake).

And I added new system notifications for promotion/demotion of guest moderators as well as the "You" detection for guest moderators.

nickvergessen commented:
Should be ready to review now?

danxuliu commented:
I have removed the fix for the avatars, because it broke the contacts menu and the avatars in messages being composed. As I already said I will fix the issue with the avatars in a different pull request, because it is not related to the changes in this one.

I have also squashed the fixes for the first commit in that commit.

Besides that I found a bug with the participant menu (the toggle is not hidden) when demoting a user (guests work fine because the list is added and removed altogether instead of updated); I have to investigate it further, although I guess that it is caused by not listening to changes in the participant type of the current user.

>  Should be ready to review now?

Yes; :+1: from me.
